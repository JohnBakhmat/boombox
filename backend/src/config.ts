import { Config, Effect } from "effect";

export const getConfig = Effect.gen(function* () {
	return {
		DB_URL: yield* Config.redacted("DB_URL"),
	};
});
