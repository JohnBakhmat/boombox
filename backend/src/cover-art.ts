import { Console, Context, Data, Effect, Layer } from "effect";

export class CoverartService extends Context.Tag("CoverartService")<
	CoverartService,
	{
		readonly fetchFrontByReleaseGroupId: (
			rgId: string,
		) => Effect.Effect<
			Uint8Array<ArrayBuffer>,
			InvalidReleaseGroupIdError | RateLimitError | UnknownFetchError,
			unknown
		>;
	}
>() {}

export const MusicbrainzCoverartLayer = Layer.succeed(
	CoverartService,
	CoverartService.of({
		fetchFrontByReleaseGroupId: (rgId) =>
			Effect.gen(function* () {
				const endpoint = `http://coverartarchive.org/release-group/${rgId}/front`;

				const response = yield* Effect.tryPromise({
					try: () => fetch(endpoint),
					catch: (err) =>
						new UnknownFetchError({
							message: "Unknown error",
							cause: err,
						}),
				});
				const status = response.status;

				if (status === 400 || status === 404) {
					return yield* Effect.fail(
						new InvalidReleaseGroupIdError({
							message: "This release group id is unparsable or invalid",
							cause: {
								releaseGroupId: rgId,
							},
						}),
					);
				}

				if (status === 503) {
					return yield* Effect.fail(
						new RateLimitError({
							message: "Rate limit hit",
						}),
					);
				}

				if (!response.ok) {
					return yield* Effect.fail(
						new UnknownFetchError({
							message: "Unknown Error",
						}),
					);
				}

				const file = yield* Effect.tryPromise({
					try: () => response.bytes(),
					catch: (err) =>
						new UnknownFetchError({
							message: "Unknown error",
							cause: err,
						}),
				});

				return file;
			}).pipe(),
	}),
);

// Errors
//

class RateLimitError extends Data.TaggedError("RateLimitError")<{
	message: string;
	cause?: unknown;
}> {}

class InvalidReleaseGroupIdError extends Data.TaggedError("InvalidReleaseGroupIdError")<{
	message: string;
	cause?: unknown;
}> {}

class UnknownFetchError extends Data.TaggedError("UnknownFetchError")<{
	message: string;
	cause?: unknown;
}> {}
