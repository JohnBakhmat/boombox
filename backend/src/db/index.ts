import { SqliteClient, SqliteMigrator } from "@effect/sql-sqlite-bun";

import * as SqliteDrizzle from "@effect/sql-drizzle/Sqlite";
import { Config, Effect, Layer, pipe } from "effect";
import { BunFileSystem } from "@effect/platform-bun";
import * as schema from "./schema";

//export const SqlLive = SqliteClient.layer({
//filename: process.env.DB_URL!,
//});

//const DrizzleLive = SqliteDrizzle.layer.pipe(Layer.provide(SqlLive));

const SqliteLive = SqliteClient.layerConfig({
	filename: Config.string("DB_URL"),
});

export class DatabaseLive extends Effect.Service<DatabaseLive>()("DatabaseLive", {
	dependencies: [SqliteLive],
	effect: SqliteDrizzle.make<typeof schema>({
		schema,
	}),
}) {}

export const DrizzleMigratorLive = pipe(
	SqliteMigrator.layer({
		loader: SqliteMigrator.fromFileSystem("./drizzle"),
		schemaDirectory: "./drizzle",
	}),
	Layer.provide(SqliteLive),
	Layer.provide(BunFileSystem.layer),
);
