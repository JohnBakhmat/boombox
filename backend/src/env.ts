import { Context, Data, Effect, Layer, Schema } from "effect";

class EnvError extends Data.TaggedError("EnvError")<{
	cause?: unknown;
	message: string;
}> {}

export class Env extends Context.Tag("Env")<Env, { readonly getEnv: Effect.Effect<EnvSchema, EnvError> }>() {}

const schema = Schema.Struct({
	DB_URL: Schema.NonEmptyString,
	FOLDER_PATH: Schema.NonEmptyString,
});

type EnvSchema = Schema.Schema.Type<typeof schema>;

export const EnvLive = Layer.succeed(Env, {
	getEnv: Schema.decodeUnknown(schema)(process.env).pipe(
		Effect.mapError((e) => new EnvError({ message: "Environment is not valid", cause: e })),
	),
});
