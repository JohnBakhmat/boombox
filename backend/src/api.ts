import { Console, Context, Data, Effect, Layer, ManagedRuntime } from "effect";
import { Elysia, status, t } from "elysia";
import { DatabaseLive } from "./db";
import { BunContext } from "@effect/platform-bun";
import { EnvLive } from "./env";
import { albumTable, artistTable, artistToAlbumTable, fileTable, songTable, songToArtistTable } from "./db/schema";
import { eq } from "drizzle-orm";
import { openapi } from "@elysiajs/openapi";
import type { Album, AlbumWithArtist, Artist, Song } from "./db/types";
import type { SqlError } from "@effect/sql";
import { pipe } from "effect";

class FileNotFoundError extends Data.TaggedError("FileNotFoundError")<{
	message: string;
	cause?: unknown;
}> {}
class AlbumNotFoundError extends Data.TaggedError("AlbumNotFoundError")<{
	message: string;
	cause?: unknown;
}> {}

type SongWithArtists = Song & { artists: Artist[] };

type GetAlbum = Album & { songs: Array<Omit<SongWithArtists, "albumId">> };

class ApiService extends Context.Tag("ApiService")<
	ApiService,
	{
		readonly getAlbumList: () => Effect.Effect<AlbumWithArtist[], SqlError.SqlError, DatabaseLive>;
		readonly getAlbum: (
			id: string,
		) => Effect.Effect<GetAlbum, SqlError.SqlError | AlbumNotFoundError, DatabaseLive>;
	}
>() {}

const ApiLive = Layer.effect(
	ApiService,
	Effect.gen(function* () {
		const db = yield* DatabaseLive;

		return {
			getAlbum: (id: string) =>
				Effect.gen(function* () {
					const rows = yield* db
						.select({
							album: albumTable,
							song: {
								id: songTable.id,
								fileId: songTable.fileId,
								title: songTable.title,
							},
							artist: artistTable,
						})
						.from(albumTable)
						.innerJoin(songTable, eq(songTable.albumId, albumTable.id))
						.innerJoin(songToArtistTable, eq(songToArtistTable.songId, songTable.id))
						.innerJoin(artistTable, eq(songToArtistTable.artistId, artistTable.id))
						.where(eq(albumTable.id, id));

					yield* Console.table(rows);

					const firstRow = rows.at(0);
					if (!firstRow) {
						return yield* new AlbumNotFoundError({
							message: "Can't extract album from selected rows",
							cause: rows,
						});
					}

					const songsMap = rows.reduce<Record<string, Omit<SongWithArtists, "albumId">>>((acc, row) => {
						const { song, artist } = row;

						if (!acc[song.id]) {
							acc[song.id] = {
								id: song.id,
								fileId: song.fileId,
								title: song.title,
								artists: [] as Artist[],
							};
						}

						if (artist) {
							acc[song.id].artists.push(artist);
						}

						return acc;
					}, {});

					const album = {
						id: firstRow.album.id,
						title: firstRow.album.title,
						songs: Object.values(songsMap),
					};

					return album;
				}),
			getAlbumList: () =>
				Effect.gen(function* () {
					const rows = yield* db
						.select({
							album: albumTable,
							artist: artistTable,
						})
						.from(albumTable)
						.innerJoin(artistToAlbumTable, eq(albumTable.id, artistToAlbumTable.albumId))
						.innerJoin(artistTable, eq(artistTable.id, artistToAlbumTable.artistId));

					const result = rows.reduce<Record<string, AlbumWithArtist>>((acc, cur) => {
						const albumId = cur.album.id;

						if (!acc[albumId]) {
							acc[albumId] = {
								...cur.album,
								artists: [],
							};
						}

						acc[albumId]?.artists.push(cur.artist);
						return acc;
					}, {});

					return Object.values(result);
				}),
		};
	}),
);

export function startApi() {
	new Elysia()
		.use(openapi())
		.get("/", "Hello Elysia")
		.get("/albums", () =>
			pipe(
				ApiService,
				Effect.andThen((x) => x.getAlbumList()),
				runtime.runPromise,
			),
		)
		.get("/album/:id", ({ params: { id } }) =>
			pipe(
				ApiService,
				Effect.andThen((x) => x.getAlbum(id)),
				runtime.runPromise,
			),
		)
		.get(
			"/file/:id",
			({ params: { id }, set }) =>
				runtime.runPromise(
					Effect.gen(function* () {
						const db = yield* DatabaseLive;

						const files = yield* db.select().from(fileTable).where(eq(fileTable.id, id)).limit(1);

						const file = files.at(0);

						if (!file) {
							return yield* Effect.fail(
								new FileNotFoundError({
									message: `Failed to find file with id ${id}`,
									cause: { id },
								}),
							);
						}

						const fileHandle = Bun.file(file.path);

						const fileExists = yield* Effect.tryPromise(() => fileHandle.exists());

						// Check if file exists
						if (!fileExists) {
							return yield* Effect.fail(
								new FileNotFoundError({
									message: `File not found on disk: ${file.path}`,
									cause: { id, path: file.path },
								}),
							);
						}

						const filename = file.path.split("/").pop();
						const encodedFilename = encodeURIComponent(filename ?? "");

						set.headers["Content-Disposition"] =
							`attachment; filename="${encodedFilename}"; filename*=UTF-8''${encodedFilename}`;
						set.headers["Content-Type"] = "audio/flac";
						set.headers["Accept-Ranges"] = "bytes";

						return fileHandle;

						return file;
					}).pipe(Effect.catchTag("FileNotFoundError", (e) => Effect.succeed(status(404, e.message)))),
				),
			{
				params: t.Object({
					id: t.String(),
				}),
			},
		)
		.get("/songs", () =>
			runtime.runPromise(
				Effect.gen(function* () {
					const db = yield* DatabaseLive;
					const rows = yield* Effect.tryPromise(() =>
						db
							.select({
								song: songTable,
								album: albumTable,
								artist: artistTable,
							})
							.from(songTable)
							.innerJoin(albumTable, eq(songTable.albumId, albumTable.id))
							.innerJoin(songToArtistTable, eq(songToArtistTable.songId, songTable.id))
							.innerJoin(artistTable, eq(songToArtistTable.artistId, artistTable.id))
							.all(),
					);

					yield* Console.dir(rows);

					const reduced = rows.reduce<Record<string, GetSongType>>((acc, row) => {
						const { song, album, artist } = row;

						if (!acc[song.id]) {
							acc[song.id] = {
								id: song.id,
								fileId: song.fileId,
								title: song.title,
								album: {
									id: album.id,
									title: album.title,
								},
								artists: [],
							};
						}

						if (artist) {
							acc[song.id].artists.push(artist);
						}

						return acc;
					}, {});

					return Object.values(reduced);
				}).pipe(Effect.withSpan("/get-songs")),
			),
		)
		.listen(3003);
}

type GetSongType = {
	id: string;
	fileId: string;
	title: string;
	album: {
		id: string;
		title: string;
	};
	artists: Array<{
		id: string;
		name: string;
	}>;
};

const layers = Layer.mergeAll(
	BunContext.layer,
	EnvLive,
	DatabaseLive.Default,
	Layer.provide(ApiLive, DatabaseLive.Default),
);
const runtime = ManagedRuntime.make(layers);
