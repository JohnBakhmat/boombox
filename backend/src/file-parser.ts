import { Env } from "./env";
import * as flac from "./flac";
import { Effect, Option, Console } from "effect";
import { FileSystem, Path } from "@effect/platform";

const SUPPORTED_EXTENSIONS = ["flac"] as const;
type SupportedExtension = (typeof SUPPORTED_EXTENSIONS)[number];

export function parseFile(path: string) {
	return Effect.gen(function* () {
		const extension = path.split(".").pop();
		const assumedType =
			extension && SUPPORTED_EXTENSIONS.includes(extension as SupportedExtension)
				? (extension as SupportedExtension)
				: null;

		if (!assumedType) {
			return yield* Effect.fail(new Error("File is unsupported"));
		}

		if (assumedType === "flac") {
			const isFlac = yield* flac.isFlac(path);
			if (isFlac) {
				const metadata = yield* flac.readMetadata(path);
				return metadata;
			}
		}
		return yield* Effect.fail(new Error("File is unsupported"));
	}).pipe(Effect.withSpan("parseFile"));
}

export function parseManyFiles(paths: string[]) {
	return Effect.gen(function* () {
		const tasks = paths.map((path) => Effect.option(parseFile(path)));

		const files = yield* Effect.all(tasks, {
			concurrency: 10,
		});

		const successful = files.filter(Option.isSome).map((file) => file.value);

		return successful;
	}).pipe(Effect.withSpan("parseManyFiles"));
}

export function readDirectory(dirPath: string, skip: string[] = []) {
	return Effect.gen(function* () {
		const fs = yield* FileSystem.FileSystem;
		const path = yield* Path.Path;

		const files = yield* fs.readDirectory(dirPath, {
			recursive: true,
		});

		const relative = files.map((file) => path.resolve(dirPath, file));
		yield* Console.log(relative);

		const filtered = relative.filter((file) => skip.includes(file) === false);
		yield* Console.log("FILTERED", filtered);

		if (filtered.length === 0) {
			return yield* Effect.succeed([]);
		}

		return yield* parseManyFiles(filtered);
	}).pipe(Effect.withSpan("readDirectory"));
}
