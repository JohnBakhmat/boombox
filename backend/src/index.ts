import { BunContext, BunRuntime } from "@effect/platform-bun";
import { Effect, Layer } from "effect";
import { DatabaseLive } from "./db";
import { Env, EnvLive } from "./env";
import { syncLibraryStream } from "./sync-library";
import { OtelLive } from "./otel";
import { startApi } from "./api";

const layers = Layer.mergeAll(BunContext.layer, EnvLive, DatabaseLive.Default);

const main = Effect.gen(function* () {
	const env = yield* Env.pipe(Effect.flatMap((x) => x.getEnv)).pipe(Effect.withSpan("env"));

	yield* syncLibraryStream(env.FOLDER_PATH);

	startApi();
}).pipe(Effect.provide(OtelLive), Effect.provide(layers));

BunRuntime.runMain(main);
