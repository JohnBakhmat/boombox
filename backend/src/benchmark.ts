import { BunContext, BunRuntime } from "@effect/platform-bun";
import { Console, Duration, Effect, Layer } from "effect";
import { DatabaseLive } from "./db";
import { Env, EnvLive } from "./env";
import { OtelLive } from "./otel";
import { readDirectory } from "./file-parser";

// Helpers
function formatBytes(bytes: number): string {
	return `${(bytes / 1024 / 1024).toFixed(2)} MB`;
}

function printMemorySnapshot(timestamp: number, heapUsed: number, heapMax: number) {
	const maxWidth = 100;

	const bar = "=".repeat((heapUsed / heapMax) * maxWidth);

	return `${bar} ${timestamp}ms: ${formatBytes(heapUsed)}\n`;
}

function printMemorySnapshots(
	iteration: number,
	memorySnapshots: Array<{ timestamp: number; rss: number; heapUsed: number }>,
) {
	return Effect.gen(function* () {
		yield* Console.log(`=== MEMORY SNAPSHOTS BY ITERATION ${iteration} ===`);

		const fileName = `./benchmark/mem-snapshots-${iteration}-${Date.now()}.txt`;

		const file = Bun.file(fileName);
		const writer = file.writer();

		const maxHeap = memorySnapshots.reduce((acc, cur) => Math.max(acc, cur.heapUsed), 0);

		for (const snapshot of memorySnapshots) {
			const line = printMemorySnapshot(snapshot.timestamp, snapshot.heapUsed, maxHeap);
			writer.write(line);
		}
		yield* Console.log("==========================");
	});
}

// ##

const layers = Layer.mergeAll(BunContext.layer, EnvLive, DatabaseLive.Default);

const iteration = (i: number) =>
	Effect.gen(function* () {
		const env = yield* Env.pipe(Effect.flatMap((x) => x.getEnv)).pipe(Effect.withSpan("env"));

		const results = {
			elapsed: 0,
			files: 0,
		};

		const memorySnapshots: Array<{ timestamp: number; rss: number; heapUsed: number }> = [];
		let peakMemory = 0;
		const startMem = process.memoryUsage();
		const startTime = Date.now();

		const interval = setInterval(() => {
			const mem = process.memoryUsage();
			const elapsed = Date.now() - startTime;

			memorySnapshots.push({
				timestamp: elapsed,
				rss: mem.rss,
				heapUsed: mem.heapUsed,
			});

			peakMemory = Math.max(peakMemory, mem.heapUsed);
		}, 50);

		const dirContent = yield* readDirectory(env.FOLDER_PATH, []).pipe(
			Effect.timed,
			Effect.tap(([duration, result]) =>
				Console.log(`Iteration ${i} took ${Duration.toMillis(duration)}ms and found ${result.length} files`),
			),
			Effect.tap(([duration, result]) => {
				results.elapsed += Duration.toMillis(duration);
				results.files += result.length;
			}),
			Effect.map(([_, result]) => result),
		);

		clearInterval(interval);
		const endMem = process.memoryUsage();

		return {
			...results,
			startTime,
			endTime: Date.now(),
			peakMemory,
			memorySnapshots,
			startMem,
			endMem,
		};
	});

type IterationResult = Effect.Effect.Success<ReturnType<typeof iteration>>;

const benchmark = Effect.gen(function* () {
	const results: Array<IterationResult> = [];
	for (let i = 0; i < 10; i++) {
		const iterationResults = yield* iteration(i);
		results.push(iterationResults);
	}

	yield* Console.dir(results);
	yield* Effect.tryPromise(() => Bun.write("./benchmark/results.json", JSON.stringify(results, null, 2)));

	const avgTime = results.reduce((acc, cur) => acc + cur.elapsed, 0) / results.length;
	const avgPeak = results.reduce((acc, cur) => acc + cur.peakMemory, 0) / results.length;
	const avgFiles = results.reduce((acc, cur) => acc + cur.files, 0) / results.length;
	const maxFiles = results.reduce((acc, cur) => Math.max(acc, cur.files), 0);
	const fileErrorRate = avgFiles / maxFiles;

	const avgHeapDelta =
		results.reduce((acc, cur) => acc + cur.endMem.heapUsed - cur.startMem.heapUsed, 0) / results.length;

	const template = `
    === BENCHMARK RESULTS ===
    Average time: ${avgTime}ms
    Files found: ${maxFiles} error rate ${fileErrorRate}
    Average peak memory: ${formatBytes(avgPeak)}
    Average heap delta: ${formatBytes(avgHeapDelta)}
    =========================
    `;

	yield* Console.log(template);

	for (const [i, result] of results.entries()) {
		yield* printMemorySnapshots(i, result.memorySnapshots);
	}
}).pipe(Effect.provide(OtelLive), Effect.provide(layers));

// @ts-expect-error - BunRuntime.runMain is not typed
BunRuntime.runMain(benchmark);
