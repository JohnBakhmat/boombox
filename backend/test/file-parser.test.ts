import { expect, it as test } from "@effect/vitest";
import { Effect } from "effect/index";
import { BunContext } from "@effect/platform-bun/index";
import { parseFile, parseManyFiles } from "../src/file-parser";
import { Console } from "effect";

const provide = Effect.provide(BunContext.layer);

test.effect("parseFile should parse flac just fine", () =>
	Effect.gen(function* () {
		const input = "./test-data/11 - Tropical Fish.flac";
		const actual = yield* parseFile(input);
		expect(actual).toHaveProperty("artist", "濱田金吾");
		expect(actual).toHaveProperty("album", "「midnight cruisin’」+「MUGSHOT」");
		expect(actual).toHaveProperty("title", "TROPICAL FISH");
	}).pipe(provide),
);

test.effect("parseManyFiles should parse many files just fine", () =>
	Effect.gen(function* () {
		const input = [
			"./test-data/11 - Tropical Fish.flac",
			"./test-data/01 - hover.mp3",
			"./test-data/01 - hover.fake.flac",
		];
		const actual = yield* parseManyFiles(input);

		yield* Console.dir(actual);

		expect(actual.length).toBe(1); // mp3 is not yet supported
	}).pipe(provide),
);
