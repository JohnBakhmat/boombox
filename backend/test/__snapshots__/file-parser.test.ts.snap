// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseFile should parse flac just fine 1`] = `
{
  "album": "「midnight cruisin’」+「MUGSHOT」",
  "artist": "濱田金吾",
  "title": "TROPICAL FISH",
}
`;
